### PR TITLE
Add copy button to toast notification

### LINKED ISSUE #11067: Add copy button to toast notifications
### Describe the project you are working on

An editor plugin that will utilize toast notifications if https://github.com/godotengine/godot/pull/98680 is merged

### Describe the problem or limitation you are having in your project

If https://github.com/godotengine/godot/pull/98680 is merged, the toast notifications that are created in editor (with `push_toast` method) are not displayed in the console for a user to easily copy, as well as the toast notification label not being selectable.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

Add a copy button to the toast notification

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

https://github.com/user-attachments/assets/f163fea2-2680-4ea2-891d-bc1561b189d4

### If this enhancement will not be used often, can it be worked around with a few lines of script?

N/A

### Is there a reason why this should be core and not an add-on in the asset library?

Should be included with the toast notifications

### PR DESCRIPTION
**Pull request by jaydensipe**
Implements and closes https://github.com/godotengine/godot-proposals/issues/11067 *Bugsquad edit: use the GitHub magic words to link the proposal and auto-close when merged :)*

# Preview (edited)
https://github.com/user-attachments/assets/04584f76-cbd1-4abe-85c4-706b3dcd2fbe

### CODE REVIEW

File: editor/gui/editor_toaster.cpp [hunk 1]
```diff
@@ -371,6 +371,20 @@ Control *EditorToaster::popup(Control *p_control, Severity p_severity, double p_
 	p_control->set_h_size_flags(SIZE_EXPAND_FILL);
 	hbox_container->add_child(p_control);
 
+	// Copy button.
+	if (p_time > 0.0) {
```
Comment (KoBeWi) [inline editor/gui/editor_toaster.cpp:375]: This condition is the same as the one below, so you can create both buttons in one block. Also the comments are unnecessary.

File: editor/gui/editor_toaster.cpp [hunk 2]
```diff
@@ -371,6 +371,20 @@ Control *EditorToaster::popup(Control *p_control, Severity p_severity, double p_
...
+		Button *copy_button = memnew(Button);
+		copy_button->set_flat(true);
+		copy_button->set_button_icon(get_editor_theme_icon(SNAME("ActionCopy")));
+		copy_button->set_focus_mode(FOCUS_NONE);
+		copy_button->set_disabled(true);
+		copy_button->connect(SceneStringName(mouse_entered), callable_mp(this, &EditorToaster::_toggle_copy_button).bind(copy_button, false));
+		copy_button->connect(SceneStringName(mouse_exited), callable_mp(this, &EditorToaster::_toggle_copy_button).bind(copy_button, true));
+		copy_button->connect(SceneStringName(pressed), callable_mp(this, &EditorToaster::copy).bind(panel));
+		copy_button->connect(SceneStringName(theme_changed), callable_mp(this, &EditorToaster::_copy_button_theme_changed).bind(copy_button));
```
Comment (KoBeWi) [inline editor/gui/editor_toaster.cpp:384]: You can unify theme changeds in one method and bind both buttons.

File: editor/gui/editor_toaster.cpp [hunk 3]
```diff
@@ -371,6 +371,20 @@ Control *EditorToaster::popup(Control *p_control, Severity p_severity, double p_
 	p_control->set_h_size_flags(SIZE_EXPAND_FILL);
 	hbox_container->add_child(p_control);
 
+	// Copy button.
+	if (p_time > 0.0) {
+		Button *copy_button = memnew(Button);
+		copy_button->set_flat(true);
+		copy_button->set_button_icon(get_editor_theme_icon(SNAME("ActionCopy")));
```
Comment (KoBeWi) [inline editor/gui/editor_toaster.cpp:378]: This is done in theme changed when button is added to tree, so unnecessary to set it here.

File: editor/gui/editor_toaster.cpp [hunk 4]
```diff
@@ -371,6 +371,20 @@ Control *EditorToaster::popup(Control *p_control, Severity p_severity, double p_
...
+		copy_button->set_focus_mode(FOCUS_NONE);
+		copy_button->set_disabled(true);
+		copy_button->connect(SceneStringName(mouse_entered), callable_mp(this, &EditorToaster::_toggle_copy_button).bind(copy_button, false));
+		copy_button->connect(SceneStringName(mouse_exited), callable_mp(this, &EditorToaster::_toggle_copy_button).bind(copy_button, true));
```
Comment (KoBeWi) [inline editor/gui/editor_toaster.cpp:382]: What's this for?

File: editor/gui/editor_toaster.cpp (hunk 2 above)
Comment (KoBeWi) [reply editor/gui/editor_toaster.cpp:384]: I opened #98794

File: editor/gui/editor_toaster.cpp (hunk 4 above)
Comment (jaydensipe) [reply editor/gui/editor_toaster.cpp:382]: So I have the button disabled by default, and it re-enables when hovered (_or_ we can change it to when the panel is hovered). I thought this looked a little less noisy then having both buttons enabled, but I'm down to change to whatever.

https://github.com/user-attachments/assets/3486003a-e3cd-475d-b2c8-0a0481147722

**VS.**

https://github.com/user-attachments/assets/0012c637-73a4-483a-8221-5606d48dc196

File: editor/gui/editor_toaster.cpp (hunk 1 above)
Comment (jaydensipe) [reply editor/gui/editor_toaster.cpp:375]: Done, ty.

File: editor/gui/editor_toaster.cpp (hunk 4 above)
Comment (KoBeWi) [reply editor/gui/editor_toaster.cpp:381]: I was wondering why the button is gray. I think they should both be enabled.

Comment (jaydensipe) [reply editor/gui/editor_toaster.cpp:381]: Fair, I'll change it rn.

File: editor/gui/editor_toaster.cpp (hunk 2 above)
Comment (jaydensipe) [reply editor/gui/editor_toaster.cpp:380]: Awesome, thanks dude. I'll change stuff when that is merged.

Comment (KoBeWi) [reply editor/gui/editor_toaster.cpp:380]: It was merged.

Comment (jaydensipe) [reply editor/gui/editor_toaster.cpp]: Rebased and updated, thanks again.

File: editor/gui/editor_toaster.cpp [hunk 5]
```diff
@@ -371,13 +371,20 @@ Control *EditorToaster::popup(Control *p_control, Severity p_severity, double p_
 	p_control->set_h_size_flags(SIZE_EXPAND_FILL);
 	hbox_container->add_child(p_control);
 
-	// Close button.
+	// Add buttons.
 	if (p_time > 0.0) {
+		Button *copy_button = memnew(Button);
+		copy_button->set_flat(true);
+		copy_button->set_focus_mode(FOCUS_NONE);
```
Comment (KoBeWi) [inline editor/gui/editor_toaster.cpp:378]: That's inconsistent with the other button 🤔

**Review by KoBeWi — APPROVED** [review]

File: editor/gui/editor_toaster.cpp (hunk 5 above)
Comment (jaydensipe) [reply editor/gui/editor_toaster.cpp]: Changed, thought it looked a bit weird after copying with the focus outline, but it's probably better for consistency's sake.

**Comment by jaydensipe** [comment]
Had a thought after watching passivestar's twitter video about the toasts, he's using them for quick tool notifications instead of errors, do we think it might be a good idea to only show the copy button if the toast has a severity of warning/error? @KoBeWi @passivestar thoughts? Or have a way to specify if you want your `push_toast` to include a copy button?

image

**Comment by passivestar** [comment]
@jaydensipe There are probably some cases where you'd want the copy button in a regular message notifications if there's a script that shows you some interesting info like some scene stats and you don't want to outright add `DisplayServer.clipboard_set()` to it yourself. It's probably not something you'd want often though

What I think might be interesting to consider in the future though is persistent toasts with action buttons:

**Comment by jaydensipe** [comment]
@passivestar Awesome, yeah that makes sense not to bloat the API for that, thanks for the input man. And yes buttons on the toasts would be sick

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**